proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,1864 @@
+the language. Salient use cases from the paper include:
+
+- Safe mutable arrays with a safe *non-copying* `freeze` operation.
+- Off-heap memory that enables allocating, reading, writing and
+  freeing memory safely, without use-after-free or double-free errors.
+  This is an important use case for latency sensitive systems
+  programming, where moving objects off-heap, out of the purview of
+  the GC, is beneficial for avoiding long GC pauses and achieving
+  predictable latencies. A prototype is implemented in the
+  `linear-base library
+  <https://github.com/tweag/linear-base/blob/master/src/Foreign/Marshal/Pure.hs>`_.
+- Safe zero-copy data (de)serialization, a notoriously difficult
+  endeavour that is in fact so error prone without linear types that
+  most production systems today typically avoid it.
+- Safe and prompt handling of system resources like files, sockets,
+  database handles etc. A `blog post
+  <http://www.tweag.io/posts/2017-08-03-linear-typestates.html>`_
+  demonstrates this use case in more detail, including tracking the
+  state of sockets in types.
+- Statically enforced communication protocols between distributed
+  processes communicating via RPC.
+
+The keyword in the above examples is **safety**. This proposal is not
+about improving the performance of the compiler's generated code. It
+is not about new runtime support. It is about enabling programmers to
+build safer API's that enforce stronger properties, thereby bringing
+*possible* but otherwise high-risk optimization techniques, like
+managing memory manually, into the realm of the *feasible*.
+
+Resource-safety or any other property are *not* an inherent property
+of linear types. They are properties of API's making careful use of
+linear types.
+
+The use cases put forth above are diverse and pervasive. Yet they are
+but a few examples of the safety properties that can be conveniently
+captured with linear types. Here are a few more:
+
+- @gelisam designed `a linear API
+  <https://github.com/gelisam/linear-examples>`_ for `3d-printable
+  models
+  <https://www.spiria.com/en/blog/desktop-software/making-non-manifold-models-unrepresentable>`_.
+- @facundominguez `shows how linear types
+  <http://www.tweag.io/posts/2017-11-29-linear-jvm.html>`_ make it
+  possible to safely manage two GC heaps managed by two separate GC's,
+  but shared between two language runtimes.
+
+Proposed Change Specification
+-----------------------------
+
+.. _Specification:
+
+We introduce a new language extension. Types with a linearity
+specification are syntactically legal anywhere in a module if and only
+if ``-XLinearTypes`` is turned on.
+
+This proposal only introduces a new type for functions. It does not
+affect the runtime system, optimization passes or code generation.
+
+Definition
+~~~~~~~~~~
+
+We say that a function ``f`` is *linear* when ``f u`` is consumed
+exactly once implies that ``u`` is *consumed exactly once* (defined
```

Comment:
repetition of "consumed exactly once"

Reply:
Sorry. I must have missed something. But I don't see a repetition here.

Reply:
NM, I misread it.